README.md
# rCorrSieve  
Michael G. Campana  
Smithsonian's National Zoo & Conservation Biology Institute  

R implementation of CorrSieve  

## Licensing  
Original R source code (*CorrSieve* versions <= 1.6-8) copyright (c) Michael G. Campana, 2010-2013 is licensed under the GNU General Public License (version 3 or later). See included LICENSE file for details.  

Public domain updates by Michael G. Campana (2019, 2022) to the GitHub documentation (*CorrSieve* versions >= 1.6-8) and package metadata (*CorrSieve* version 1.6-9) are United States government works.  

## Introduction
*CorrSieve* is a Ruby and R package that filters *Q* value output from the programs STRUCTURE (Pritchard et al. 2000) and INSTRUCT (Gao et al. 2007) by correlation values. It outputs matrices showing significant correlations between individual runs for each *K*. It can also calculate Δ*K* (Evanno et al. 2005), mean *F*<sub>ST</sub>s and Δ*F*<sub>ST</sub>. These measures help identify meaningful values of *K*.  

## Installation and Compatibility  
rCorrSieve is compatible with Windows, Linux, and UNIX (including macOS) operating systems. It requires the R statistical environment available from the Comprehensive R Archive Network (CRAN: http://cran.r-project.org). Install the appropriate environment for your operating system.  

Load the R environment.  

The latest stable rCorrSieve version is available from CRAN. Install this version using:  
`install.packages('corrsieve')`  

The development rCorrSieve version is hosted here on GitHub. Installation of this version requires the `remotes` package:  
`install.packages('remotes')`  
`library(remotes)`  
`install_github('campanam/rCorrSieve')`  

## Usage  
1. Prepare input for *CorrSieve*. *CorrSieve* reads directly from STRUCTURE and INSTRUCT output files, but requires that all files be in a single folder. Do not place other files in this folder. All files should end in ‘_f’ without an extension, e.g:  
TEST_11_f  
TEST_12_f  
TEST_13_f  
TEST_21_f  
TEST_22_f  
TEST_23_f  

2. Load the package:  
`library(corrsieve)`  

3. The command to read a folder of STRUCTURE output for summarising *F*<sub>ST</sub> and ln P(D) and calculating Δ*K* and Δ*F*<sub>ST</sub> is `read.struct(filepath = “./”, instruct = FALSE)` The file path default is the current directory. The default data file format is the STRUCTURE format. Setting `instruct` to TRUE indicates the presence of INSTRUCT data. This will create a table of data in the R workspace. For example, the command:  
`data <- read.struct(“C:/Results/”)`  
will create the table ‘data’ containing all the STRUCTURE output in the folder `C:/Results/`.  

4. To summarise the *F*<sub>ST</sub> and ln P(D) data from the data read with `read.struct`, the commands `summarise.Fst` and `summarise.lnPD` are used respectively. For example, using the data we read in previously, the commands:  
`lnpd <- summarise.lnPD(data)`  
`fst <- summarise.Fst(data)`  
will create the tables ‘lnpd’ and ‘fst’ containing the summarised ln P(D) and *F*<sub>ST</sub> data respectively.  

Since the assigned clusters will not necessarily be in the same order for each STRUCTURE/ INSTRUCT run, `summarise.Fst` has a second argument `stdevopt` to best order the *F*<sub>ST</sub> data. Setting `stdevopt = 1` (the default) uses no optimisation procedure, setting `stdevopt = 2` orders the *F*<sub>ST</sub> data by value, whilst setting `stdevopt = 3` will use correlation matrices between runs to determine the optimal *F*<sub>ST</sub> data ordering. Using the correlation matrices will prompt the user to enter the file path for the original files again.  

*NOTE:* F<sub>ST</sub> *statistics are only available from STRUCTURE data generated under the admixture model. Output generated in INSTRUCT (even under the admixture model) or under other STRUCTURE models will cause an error.*  

5. To calculate, Δ*K* and Δ*F*<sub>ST</sub> from the summarised ln P(D) and *F*<sub>ST</sub> data, the command `calc.delta(input, Fst = FALSE)` is used. The default is ln P(D) data. The argument `Fst` determines whether the input data is *F*<sub>ST</sub> or ln P(D) data. For example, the commands:  
`deltaK <- calc.delta(lnpd)`  
`deltaF <- calc.delta(fst, Fst= TRUE)`  
will create the tables ‘deltaK’ and ‘deltaF’ containing the calculated Δ*K* and Δ*F*<sub>ST</sub> statistics as well as the intermediary statistics L′(*K*), L′′(*K*), F′(*K*) and F′′(*K*) used to calculate Δ*K* and Δ*FST* (Campana et al. 2011; Evanno et al. 2005). These statistics are denoted by LprK, LdprK, FprK, and FdprK respectively.  

6. To perform *Q*-matrix correlations, the command`corr.Qmatrix(filepath = "./", rowncol = TRUE, avmax = TRUE, pvalue = FALSE, raw = TRUE, r = 0.99, p = 0.05)` is used. The file path default is the current directory. The default data file format is the STRUCTURE format. Setting `instruct` to TRUE indicates the presence of INSTRUCT data. The arguments `rowncol`, `avmax`, and `pvalue` determine whether the rows-and-columns criterion (Campana et al. 2011), the average max correlation criterion (Cockram et al. 2008), permutation tests are used. The argument `raw` determines whether unfiltered *Q* matrix correlation matrices are outputted. The arguments `r` and `p` determine the minimum Pearson coefficient and the maximum *p* values to be considered a significant correlation. The data generated by `corr.Qmatrix` are entered into an S4 object of class ‘QmatrixFilt’. For example:  
`Qmat <- corr.Qmatrix(“C:/Results/”, pvalue = TRUE)`  
will generate an S4 object named ‘Qmat’ in which the STRUCTURE output in the folder `C:/Results/` are correlated and filtered both by the average maximum correlation and the rows-and-columns criteria. Permutation tests are also used. The unfiltered correlation matrices are outputted. The minimum Pearson coefficient is 0.99 and the maximum *p* value is 0.05.

7. To view, the various correlation matrix outputs, the commands `@rowncol`, `@avmaxcorr` and `@rawcorr` are used. For example:
`Qmat@rowncol` will show the output of the rows-and-columns criterion  
`Qmat@avmaxcorr` will show the output of the average maximum correlation criterion  
`Qmat@rawcorr` will show the unfiltered correlation matrices.  

## Bugs and Contributing
Please report all bugs (and any suggestions for improvements) to Michael G. Campana (campanam@si.edu).  

## CorrSieve Citation  
Campana et al. 2011. *CorrSieve*: software for summarizing and evaluating Structure output. Mol. Ecol. Resour. 11:349-352. https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1755-0998.2010.02917.x  

## Acknowledgments
Rita Cannas helpfully checked the method for calculating Δ*K*. Dent Earl , Elena López, and Michał Żmihorski identified bugs in the software.  

## References
Campana et al. 2011. *CorrSieve*: software for summarizing and evaluating Structure output. Mol. Ecol. Resour. 11:349-352. https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1755-0998.2010.02917.x  

Cockram et al. 2008. Association mapping of partitioning loci in barley. BMC Genet. 9: 16–29. https://bmcgenet.biomedcentral.com/articles/10.1186/1471-2156-9-16.  

Evanno et al. 2005. Detecting the number of clusters of individuals using the software STRUCTURE: a simulation study. Mol. Ecol. 14: 2611-2620. https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1365-294X.2005.02553.x.  

Gao et al. 2007. A Markov Chain Monte Carlo approach for joint inference of population structure and inbreeding rates from multilocus genotype data. Genetics. 176: 1635-1651. https://www.genetics.org/content/176/3/1635.  

Hester et al. 2019. remotes: R package installation from remote repositories, including 'GitHub'. R package version 2.1.0. https://CRAN.R-project.org/package=remotes.  

Pritchard et al. 2000. Inference of population structure using multilocus genotype data. Genetics. 155: 945–49. https://www.genetics.org/content/155/2/945.  


CHANGELOG.md
# rCorrSieve Change Log

Michael G. Campana, 2019-2022  
Smithsonian's National Zoo & Conservation Biology Institute  
Contact: campanam@si.edu  

Public domain rCorrSieve GitHub documentation updates (version = 1.6-8) are US government works.  
Unmodified *CorrSieve* source code (versions <= 1.6-8) copyright (c) Michael G. Campana, 2010-2013

### Version 1.6-9  
CRAN Metadata update  

### Version 1.6-8 (GitHub 2019)  
Updated documentation and licensing for GitHub hosting  

### Version 1.6-8  
CRAN Metadata update  

### Version 1.6-7  
CRAN Metadata update  

### Version 1.6-6  
Fixed a bug in calc.delta that prevented calculations when K = 1 was not included.  

### Version 1.6-5  
Added compatibility with INSTRUCT output.  
Fixed bug in which read.struct assumed the presence of *F*<sub>ST</sub> values, even in non-admixture model runs.  
Fixed bug in the *F*<sub>ST</sub> optimisation procedure using ordering of *F*<sub>ST</sub> values.  
Improved algorithm for *F*<sub>ST</sub> optimisation using *Q*-matrix correlations.  

### Version 1.6-4  
Corrected bug which prevented rows-and-columns *Q*-matrix correlations from being calculated when only two runs were completed per *K*.  
Corrected bug which prevented the performance of *Q*-matrix correlations when the number of runs per *K* varied between *K* values or where the minimum tested *K* value was not 1.  
Corrected bug which prevented the calculation of the average maximum correlation when the number of runs for individual *K* values was 0.  

### Version 1.6-3  
Corrected minor bug in testing algorithm for continuous data  
Corrected minor bug in setting default file location  
Documentation corrections  

### Version 1.6-2  
Added 2 methods to optimise the standard deviation calculation.  

### Version 1.6-1  
Translated original Ruby program to R.  

### Versions < 1.6-1  
Only available in Ruby implementation (rubyCorrSieve)  


R/corrsieve.R
#-----------------------------------------------------------------------------------------------
# CorrSieve 1.6-8: software to summarise and evaluate output from STRUCTURE
# Copyright (C) 2010-2013 Michael G. Campana
# GitHub Documentation revisions by Michael G. Campana (2019) are in the
# public domain as a US government work.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#-----------------------------------------------------------------------------------------------

setClass("matrixCorr", representation(K="numeric",Run1="numeric",Run2="numeric",CorrMatrix="matrix",Pvalues="matrix"))
matrixCorr <- function(K, Run1, Run2, CorrMatrix, Pvalues = matrix(NA)) {
	res <- new("matrixCorr")
	res@K <- K
	res@Run1 <- Run1
	res@Run2 <- Run2
	res@CorrMatrix <- CorrMatrix
	res@Pvalues <- Pvalues
	return(res)
}
setClass("rowncolMatrix", representation(K="numeric", filtermatrix = "table"))
rowncolMatrix <- function(K, filtermatrix) {
	res <- new("rowncolMatrix")
	res@K <- K
	res@filtermatrix <- filtermatrix
	return(res)
}
setClass("QmatrixFilt", representation(rowncol="list",avmaxcorr="table",rawcorr="list"))
QmatrixFilt <- function(rowncol = list(""), avmaxcorr = as.table(matrix(NA)), rawcorr = list("")) {
	res <- new("QmatrixFilt")
	res@rowncol <- rowncol
	res@avmaxcorr <- avmaxcorr
	res@rawcorr <- rawcorr
	return(res)
}
read.struct <- function(filepath = "./", instruct = FALSE) {
	options(warn = -9)
	files <- list.files(path = filepath, pattern = "_f$")
	maxK <- 1
	for (i in files) {
		lin <- paste(filepath, i, sep = "")
		x <- readLines(lin)
		if (instruct == TRUE) {
			K <- x[c(grep("Population number assumed=", as.character(x), perl = TRUE))]
			K <- as.numeric(substring(K, 31, nchar(K)))}
		else {
			K <- x[c(grep("populations assumed", as.character(x), perl = TRUE))]
			K <- as.numeric(substring(K, 4, nchar(K)-20))}
		if (K > maxK) {maxK <- K}
	}
	data <- "NULL"
	for (i in files) {
		lin <- paste(filepath, i, sep = "")
		x <- readLines(lin)
		if (instruct == TRUE) {
			K <- x[c(grep("Population number assumed=", as.character(x), perl = TRUE))]
			lnPD <- x[c(grep("Posterior Mean = ", as.character(x), perl = TRUE))]
			K <- as.numeric(substring(K, 31, nchar(K)))}
		else {
			K <- x[c(grep("populations assumed", as.character(x), perl = TRUE))]
			lnPD <- x[c(grep("Estimated Ln Prob of Data", as.character(x), perl = TRUE))]
			K <- as.numeric(substring(K, 4, nchar(K)-20))}
		Fsts <- c()
		for (j in 1:K) {
			line <- paste("Mean value of Fst_", as.character(j), " ", sep = "")
			Fst <- x[c(grep(line, as.character(x), perl = TRUE))]
			Fst <- as.numeric(substring(Fst, nchar(Fst)-6, nchar(Fst)))
			Fsts <- c(Fsts, Fst)
		}
		if (length(Fsts) < maxK) {
			for (l in 1:(maxK-length(Fsts))) {
				Fsts <- c(Fsts, NA)
			}
		}
		if (instruct == TRUE) {
			lnPD <-as.numeric(substring(lnPD, 22, 31))}
		else {
			lnPD <- as.numeric(substring(lnPD, 31, 37))}
		dat <- matrix(c(K, lnPD, Fsts), ncol = (maxK + 2),byrow = TRUE)
		if (is(data, "character")) {data <- dat} else {data <- rbind(data, dat)}
	}
	nam <- c("K", "lnPD")
	for (i in 1:maxK) {
		nam <- c(nam, paste("Fst_",as.character(i),sep= ""))
	}
	colnames(data) <- nam
	data <- data.frame(data)
	data <- data[order(data$K),]
	return(data)
}
corr.Qmatrix <- function(filepath  = "./", instruct = FALSE, rowncol = TRUE, avmax = TRUE, pvalue = FALSE, raw = TRUE, r = 0.99, p = 0.05) {
	options(warn = -9)
	files <- list.files(path = filepath, pattern = "_f$")
	maxK <- 1
	for (i in files) {
		lin <- paste(filepath, i, sep = "")
		x <- readLines(lin)
		if (instruct == TRUE) {
			K <- x[c(grep("Population number assumed=", as.character(x), perl = TRUE))]
			K <- as.numeric(substring(K, 31, nchar(K)))}
		else {
			K <- x[c(grep("populations assumed", as.character(x), perl = TRUE))]
			K <- as.numeric(substring(K, 4, nchar(K)-20))}
		if (K > maxK) {maxK <- K}
	}
	Qmatrices <- "NULL"
	NoColK <- c()
	for (j in 1:maxK) {
		Kfiles <- c()
		for (i in files) {
			lin <- paste(filepath, i, sep = "")
			x <- readLines(lin)
			if (instruct == TRUE) {
				K <- x[c(grep("Population number assumed=", as.character(x), perl = TRUE))]
				K <- as.numeric(substring(K, 31, nchar(K)))}
			else {
				K <- x[c(grep("populations assumed", as.character(x), perl = TRUE))]
				K <- as.numeric(substring(K, 4, nchar(K)-20))}
			if (j == K) {Kfiles <- c(Kfiles, lin)}
		}
		NoColK[j] <- length(Kfiles)
		for (i in Kfiles) {
			Qmatrix <- c()
			x <- readLines(i)
			Qmat <- x[c(grep(" : ", as.character(x), perl = TRUE))]
			Qmat <- Qmat[c(grep("cluster", as.character(Qmat), invert = TRUE, perl = TRUE))]
			Qmat <- Qmat[c(grep("Cluster", as.character(Qmat), invert = TRUE, perl = TRUE))]
			Qmat <- Qmat[c(grep("Locus", as.character(Qmat), invert = TRUE, perl = TRUE))]
			Qmat <- gsub("inf", "0.000", Qmat, perl = TRUE)
			for (l in 1:j) {
				if (instruct == TRUE && l == 1) {beg <-nchar(Qmat)-4} else {beg <- nchar(Qmat)-5}
				Qmat2 <- as.numeric(substring(Qmat, beg, beg+4))
				Qmat <- substring(Qmat, 1, beg - 1)
				Qmatrix <- c(Qmat2, Qmatrix)
			}
			Qmatrix <- matrix(Qmatrix, nrow = j, byrow = TRUE)
			Qmatrix <- t(Qmatrix)
			if (is(Qmatrices, "character")) {Qmatrices <- Qmatrix} else {Qmatrices <- cbind(Qmatrices, Qmatrix)}
		}
	}
	Qmatrices <- data.frame(Qmatrices)
	rawmatrix <- c()
	avmaxcorr <- c()
	rowcolALL <- c()
	rowcolcorr <- c()
	end <- 0
	for (i in 1:maxK) {	
		start <- end + 1
		end <- end + (NoColK[i] * i)
		if (NoColK[i] > 1) {
			corr <- cor(Qmatrices[,start:end], Qmatrices[,start:end])
			if (pvalue == TRUE) {
				sub <- Qmatrices[,start:end]
                        len  <- i * NoColK[i]
                        pvar <- c()
                        for (j in 1:len) {
                                xt <- c(sub[,j])
                                for (k in 1:len) {
                                        yt <- c(sub[,k])
                                        ptmp <- cor.test(xt, yt, method = "pearson")
                                        pvar <- c(pvar, ptmp$p.value)
                                }
                        }
                        pvar <- matrix(pvar, ncol = len, byrow = TRUE)
			}
			maxcorr <- c()
			rowcolcorr <- c()
			for (j in 1:NoColK[i]) {
				ya <- (j-1)*i + 1
				yb <- j*i
				min <- j+1
				if (min < NoColK[i] + 1) {
					for (k in min:NoColK[i]) {
						xa <- (k-1)*i+1
						xb <- k*i
						rmat <- matrix(corr[ya:yb,xa:xb], ncol = i, byrow = TRUE)
						if (pvalue == TRUE) {
							pv <- matrix(pvar[ya:yb,xa:xb], ncol = i, byrow = TRUE)
							mat <- matrixCorr(K = i, Run1 = j, Run2 = k, CorrMatrix = rmat, Pvalues = pv)
							if (i > 1) {
								for (l in 1:i) {
									for (m in 1:i) {
                                        if (is.na(pv[l,m])) {rmat[l,m] <- NA} else {
                                            if (pv[l,m] > p) {rmat[l,m] <- 0}
                                        }
									}
								}
						}
						} else {mat <- matrixCorr(K = i, Run1 = j, Run2 = k, CorrMatrix = rmat)}
						rawmatrix <- c(rawmatrix, mat)
                        mx <- max(abs(rmat))
                        maxcorr <- c(maxcorr, mx)
						if (rowncol == TRUE) {
							if (i == 1) {rowcolcorr <- c(rowcolcorr, "?")} else {
								sig = TRUE
                                corflag = FALSE
								for (m in 1:i) {
                                    if (NA %in% rmat[m,]) {corflag <- TRUE} else {
                                        if (max(abs(rmat[m,])) < r) {sig <- FALSE}
                                    }
								}
								for (m in 1:i) {
                                    if (NA %in% rmat[,m]) {corflag <- TRUE} else {
                                        if (max(abs(rmat[,m])) < r) {sig <- FALSE}
                                    }
								}
								if (corflag == TRUE) {rowcolcorr <-c(rowcolcorr, "?")} else {
                                    if (sig == TRUE) {rowcolcorr <- c(rowcolcorr, "Y")} else {rowcolcorr <- c(rowcolcorr, "N")}
                                }
							}
						}
					}
				}
			}
			if (rowncol == TRUE) {
				if (NoColK[i] > 2) {
					for (z in 1:(NoColK[i]-2)) {
						ca <- rowcolcorr[1:(NoColK[i]*z-z)]
						for (y in 1:z) {ca <- c(ca, NA)}
						rowcolcorr <- c(ca, rowcolcorr[(NoColK[i]*z-z + 1):length(rowcolcorr)])
					}
				}
				rowcolcorr <- as.table(matrix(rowcolcorr, ncol = (NoColK[i] - 1), byrow = TRUE))
				rownames(rowcolcorr) <- c(2:NoColK[i])
				colnames(rowcolcorr) <- c(1:(NoColK[i]-1))
				rowcolcorr <- rowncolMatrix(K = i, filtermatrix = rowcolcorr)
				rowcolALL <- c(rowcolALL, rowcolcorr)
			}
			avmaxcorr <- c(avmaxcorr, mean(maxcorr))
		}else {
			avmaxcorr <- c(avmaxcorr, NA)
			rowcolcorr <- c(rowcolcorr, table(NA))
			}
	}
	results <- QmatrixFilt()
	if (avmax == TRUE) {
		avmaxfilter <- c()
		for (i in 1:maxK){
			if (is.na(avmaxcorr[i]) == TRUE) 
				{avmaxfilter <- c(avmaxfilter, "?")}
			else	
				{if (avmaxcorr[i] < r) {avmaxfilter <- c(avmaxfilter, "N")} else {avmaxfilter <- c(avmaxfilter, "Y")}}			
		}	
		avmaxcorr <- rbind(c(1:maxK),avmaxcorr,avmaxfilter)
		avmaxcorr <- as.table(matrix(avmaxcorr, nrow = 3, byrow = FALSE))
		rownames(avmaxcorr) <- c("K", "Average Max Corr", "Significantly Correlated")
		results@avmaxcorr <- avmaxcorr
	}
	if (raw == TRUE) {results@rawcorr <- rawmatrix}
	if (rowncol == TRUE) {results@rowncol <- rowcolALL}
	return(results)
}
summarise.lnPD <- function(input) {
	lnPD <- "NULL"
	for (i in min(input$K):max(input$K)) {
		data <- c()
		for (j in 1:nrow(input)){
			if (input$K[j]==i) {data <- c(data, input$lnPD[j])}
		}
		dat <- matrix(c(i, mean(data), sd(data)), ncol = 3, byrow = TRUE)
		if (is(lnPD, "character")) {lnPD <- dat} else {lnPD <- rbind(lnPD, dat)}
	}
	colnames(lnPD) <- c("K", "Mean", "StDev")
	lnPD <- data.frame(lnPD)
	return(lnPD)
}
summarize.lnPD <- function(input) {return(summarise.lnPD(input))}
summarise.Fst <- function(input, stdevopt = 1) {
	if (stdevopt == 3) {
		cat("\n", "Enter raw data file path", "\n")
		filepath <- scan(n=1, what = "character", quiet = TRUE)
	}
	Fst <- "NULL"
	sdmeans <- c()
	meansds <- c()
	sdssds <- c()
	fmeans <- c()
	fsds <- c()
	for (i in min(input$K):max(input$K)) {
		data <- c()
		for (j in 1:nrow(input)){			
			if (input$K[j]==i) {
				for (k in 1:i) {data <- c(data, input[,k+2][j])}	
			}
		}
		tmp <- matrix(data, ncol = i, byrow = TRUE)
		means <- c()
		sds <- c()
		if (stdevopt == 2) {
			for (kc in 1:nrow(tmp)){
				tmp2 <- t(tmp[kc,1:ncol(tmp)])
				tmp2 <- t(tmp2[order(tmp2[1,])])
				tmp[kc, 1:ncol(tmp2)] <- tmp2
			}
		}
		if (stdevopt == 3 && i > 1) {
			options(warn = -9)
			files <- list.files(path = filepath, pattern = "_f$")
			Qref <- "NULL"
			Qmatrices <- "NULL"
			Kfiles <- c()
			for (fil in files) {
				lin <- paste(filepath, fil, sep = "")
				x <- readLines(lin)
				K <- x[c(grep("populations assumed", as.character(x), perl = TRUE))]
				K <- as.numeric(substring(K, 4, nchar(K)-20))
				if (i == K) {Kfiles <- c(Kfiles, lin)}
			}
			for (fil in Kfiles) {
				Qmatrix <- c()
				x <- readLines(fil)
				Qmat <- x[c(grep(" : ", as.character(x), perl = TRUE))]
				Qmat <- Qmat[c(grep("cluster", as.character(Qmat), invert = TRUE, perl = TRUE))]
				Qmat <- Qmat[c(grep("Locus", as.character(Qmat), invert = TRUE, perl = TRUE))]
				for (l in 1:i) {
					beg <- nchar(Qmat)-5
					Qmat2 <- as.numeric(substring(Qmat, beg, beg+4))
					Qmat <- substring(Qmat, 1, beg-1)
					Qmatrix <- c(Qmat2, Qmatrix)	
				}
				
				Qmatrix <- matrix(Qmatrix, nrow = i, byrow = TRUE)
				Qmatrix <- t(Qmatrix)
				if (is(Qmatrices, "character") && is(Qref, "character")) {Qref <- Qmatrix}
					else if (is(Qmatrices, "character")) {Qmatrices <- Qmatrix} else {Qmatrices <- cbind(Qmatrices, Qmatrix)}
			}
			xb <- 0
			for (jy in 2:length(Kfiles)) {
				xa <- xb +1
				xb <- xb + i
				tmp2 <- c()
				rmat <- cor(Qref, Qmatrices[,xa:xb])
				for (ord in 1:i){
					rv <- -1.0
					sp <- 1
					for (a in 1:i) {
						if (rmat[ord,a] > rv) {
							rv <- rmat[ord,a]
							sp <- a}
					}
					tmp2 <- c(tmp2, tmp[jy,sp])
					rmat[,sp] <- -2.0
					}		
				tmp[jy,] <- tmp2
			}
		}
		for (j in 1:i) {
			means <- c(means, mean(tmp[,j]))
			sds <- c(sds, sd(tmp[,j]))
		}
		sdmeans <- c(sdmeans, sd(means))
		meansds <- c(meansds, mean(sds))
		sdssds <- c(sdssds, sd(sds))
		fmeans <- c(fmeans, means)
		fsds <- c(fsds, sds)
		if (i < max(input$K)) {
			for (j in 1:(max(input$K)-i)) {
				fmeans <- c(fmeans, NA)
				fsds <- c(fsds, NA)
			}
		}
		dat <- matrix(c(i, mean(data), sd(data)), ncol = 3, byrow = TRUE)
		if (is(Fst, "character")) {Fst <- dat} else {Fst <- rbind(Fst, dat)}
	}
	sdmeans <- matrix(sdmeans, ncol = 1, byrow = TRUE)
	meansds <- matrix(meansds, ncol = 1, byrow = TRUE)
	sdssds <- matrix(sdssds, ncol = 1, byrow = TRUE)
	fmeans <- matrix(fmeans, ncol = max(input$K), byrow = TRUE)
	fsds <- matrix(fsds, ncol = max(input$K), byrow = TRUE)
	Fst <- cbind(Fst, sdmeans, meansds, sdssds, fmeans, fsds)
	nam <- c()
	for (i in 1:max(input$K)) {
		nam <- c(nam, paste("Mean_Fst_",as.character(i),sep= ""))
	}
	for (i in 1:max(input$K)) {
		nam <- c(nam, paste("StDev_Fst_",as.character(i),sep= ""))
	}
	colnames(Fst) <- c("K", "Overall_Mean", "Overall_StDev", "StDev_of_Means", "Mean_StDev", "StDev_of_StDev", nam)
	Fst <- data.frame(Fst)
	return(Fst)
}
summarize.Fst <- function(input, stdevopt = 1) {return(summarise.Fst(input, stdevopt))}
calc.delta <- function(input, Fst = FALSE) {
	flag <- FALSE
	j <- min(input$K)
	for (i in min(input$K):(max(input$K)-1)) {
		if (i+1 != input$K[i-j+2]) {flag <- TRUE}
	}
	if (flag == TRUE) {print("Discontinuous data. Cannot process.")} else {
		LprK <- c(NA)
		LdprK <- c()
		deltaK <- c()
		for (i in 2:nrow(input)) {
			if (Fst ==FALSE) {lpr <- input$Mean[i] - input$Mean[i-1]} else {lpr <-input$Overall_Mean[i] - input$Overall_Mean[i-1]}
			LprK <- c(LprK, lpr)
		}
		for (i in 2:nrow(input)-1) {
			ldpr <- abs(LprK[i+1] - LprK[i])
			LdprK <- c(LdprK, ldpr)
			if (Fst ==FALSE) {dK <- ldpr/input$StDev[i]} else {dK <- ldpr/input$Overall_StDev[i]}
			deltaK <- c(deltaK, dK)
		}
		LdprK <- c(LdprK, NA)
		deltaK <- c(deltaK, NA)
		output <- matrix(c(input$K,LprK,LdprK,deltaK),ncol=nrow(input),byrow =TRUE)
		output <- t(output)
		if (Fst == FALSE) {colnames(output) <- c("K","LprK","LdprK","deltaK")} else {colnames(output) <- c("K","FprK","FdprK","deltaFst")}
		output <- data.frame(output)
		return(output)
	}
}
